Add a CityValidator for validating city names.

jfly opened the pull request:
(This is part of #361.)

I had some fun with Ruby here. My thinking is that the validations we
want to enforce will be a bit different depending on the country, so I
made it possible to create a `CountryCityValidator` for each
country. These `CountryCityValidator`s know how to check if a city name
is valid or not. If there is no `CountryCityValidator` for a given
country, then we treat the city name as valid. I could see us changing
that in the future once we've added a `CountryCityValidator` for every
country we host competitions in.

![image](https://user-images.githubusercontent.com/277474/45043567-5bc7d680-b022-11e8-8ea7-c87694ed5acf.png)

jonatanklosko commented on WcaOnRails/lib/city_validator.rb:
```diff
@@ -0,0 +1,61 @@
+  def reason_why_invalid(city)
+    _city, region = city.split(", ")
+    if region.nil?
+      "is not of the form 'city, #{@type_of_region}'"
```
Thought: should we think about I18n here?

jfly replied:
Ooooh, good point, I totally didn't think about that! Currently, this is only going to be used by delegates, so I have a slight preference for leaving this as is for now, and we can always internationalize it later if we decide we want to?

Another reason I'm a bit loathe to i18n it now is because I think it may change a bit as we add support for new countries, and it would be nice to save our translators some churn.

Definitely willing to i18n this now if you or @viroulep feel differently, though!

jonatanklosko replied:
We have the competition form internationalized, so it would be consistent to do so here as well, but it's not of much importance and I'm all for merging this up and seeing if it does the job as expected first :ship:

jonatanklosko commented on WcaOnRails/lib/city_validator.rb:
```diff
@@ -0,0 +1,61 @@
+  end
+
+  def self.add_validator_for_country(country_iso2, validator)
+    @validators_by_country_iso2 ||= {}
```
Doesn't matter here, but what about just initializing this class instance variable with `{}`? What I mean is `@validators_by_country_iso2 = {}` outside the methods. Theoretically you could call `self.get_validator_for_country` before `self.add_validator_for_country`, which would result in an error now.

jfly replied:
Oh, good call! That's definitely better.

jfly commented:
Updated in response to @jonatanklosko's comments. I also added all the US states and Canadian provinces, which motivated me to pull the country city validators out into their own files.

At this point, I'm fairly pleased with this. I am going to reach out to the WRT now to figure out if there's any trickiness involved in going live with this.

viroulep commented:
Just to check: are we sure all the `not_over` competitions are valid?
It would be inconvenient that when some Delegates/organizers want to update their events' rounds or schedule, the changes would be rejected because the competition is invalid :s

jfly commented:
Ah, good question, @viroulep!! I just tried running this locally with my developer database export, and discovered a bug.

# Spot the bug!
Here's how the code used to be.

**Question**: How many states does this create?

```
US_STATES = %w(
  California
  New York
).to_set
```
**Answer**: This creates 3 states: "California", "New", and "York". Eep.

I see two options to fix this:

# Option 1

Escape the space. This works, but does clutter the list a bit.

```
US_STATES = %w(
  California
  New\ York
).to_set
```

# Option 2

Create a newline separated list of states, and manually split and reject blank lines. This is nice in that it doesn't clutter the list with weird backslashes, but it does mean we have to do some potentially code-smelly string manipulation to generate our list.

```
US_STATES = " 
California
New York
".split("\n").reject(&:blank?).to_set
```

Do y'all have any preference?

jfly commented:
After fixing the bug described above, I am able to successfully validate all `cityNames` in our Competitions table (this is running against my local copy of the developer database export):

```
irb(main):004:0> Competition.all.reject { |c| c.valid? || c.errors[:cityName].length == 0 }.length
=> 0
```

I'll re-run this against the latest version of the developer database dump, and report if any issues come up!

viroulep commented on WcaOnRails/spec/lib/city_validator_spec.rb:
```diff
@@ -0,0 +1,62 @@
+      expect(model).to be_valid
+
+      model.city = "New York, NY"
+      expect(model).to be_invalid_with_errors city: ["CA is not a valid state"]
```
Ha, the state in the error hasn't been updated :p

viroulep commented on WcaOnRails/spec/lib/city_validator_spec.rb:
```diff
@@ -0,0 +1,62 @@
+      expect(model).to be_valid
+
+      model.city = "Dieppe, NB"
+      expect(model).to be_invalid_with_errors city: ["AB is not a valid province"]
```
Same here :)

jfly replied:
Eep, fixed!

jfly commented:
Haha, ok. Happy to change it. Let's see if @jonatanklosko has a preference =)

jonatanklosko commented:
Well, the other way there's a risk of someone adding a new state and forgetting about escaping a space, but given that states doesn't change often, this is probably irrelevant :smiley:

Honestly, I'm fine with either solution as long as the states are indented (I'm not in love with them being as the same level as the constant). That said, `%w()` instead of `split("\n").reject(&:blank?)` sounds neater to me =)

jfly commented:
Thanks for the feedback! I've gone with the `%w()` solution and escaped all the whitespace.

I plan to merge this up tomorrow afternoon if no one objects before then!
